src/renderer/atlas/AtlasEngine.api.cpp

```diff
@@ -0,0 +1,599 @@
+template<typename T, typename U>
+constexpr HRESULT api_narrow(U val, T& out) noexcept
+{
+    out = static_cast<T>(val);
+    return static_cast<U>(out) != val || (std::is_signed_v<T> != std::is_signed_v<U> && out < T{} != val < U{}) ? HRESULT_FROM_WIN32(ERROR_ARITHMETIC_OVERFLOW) : S_OK;
+}
+#pragma warning(pop)
+
+template<typename T, typename U>
+constexpr HRESULT vec2_narrow(U x, U y, AtlasEngine::vec2<T>& out) noexcept
+{
+    return api_narrow(x, out.x) | api_narrow(y, out.y);
+}
+
+#pragma region IRenderEngine
+
+[[nodiscard]] HRESULT AtlasEngine::Invalidate(const SMALL_RECT* const psrRegion) noexcept
+{
+    //assert(psrRegion->Top < psrRegion->Bottom && psrRegion->Top >= 0 && psrRegion->Bottom <= _api.cellCount.y);
+
+    // BeginPaint() protects against invalid out of bounds numbers.
+    _api.invalidatedRows.x = std::min(_api.invalidatedRows.x, gsl::narrow_cast<u16>(psrRegion->Top));
+    _api.invalidatedRows.y = std::max(_api.invalidatedRows.y, gsl::narrow_cast<u16>(psrRegion->Bottom));
+    return S_OK;
+}
+
+[[nodiscard]] HRESULT AtlasEngine::InvalidateCursor(const SMALL_RECT* const psrRegion) noexcept
+{
+    //assert(psrRegion->Left <= psrRegion->Right && psrRegion->Left >= 0 && psrRegion->Right <= _api.cellCount.x);
+    //assert(psrRegion->Top <= psrRegion->Bottom && psrRegion->Top >= 0 && psrRegion->Bottom <= _api.cellCount.y);
+
+    const auto left = gsl::narrow_cast<u16>(psrRegion->Left);
+    const auto top = gsl::narrow_cast<u16>(psrRegion->Top);
+    const auto right = gsl::narrow_cast<u16>(psrRegion->Right);
+    const auto bottom = gsl::narrow_cast<u16>(psrRegion->Bottom);
+
+    // BeginPaint() protects against invalid out of bounds numbers.
+    _api.invalidatedCursorArea.left = std::min(_api.invalidatedCursorArea.left, left);
+    _api.invalidatedCursorArea.top = std::min(_api.invalidatedCursorArea.top, top);
+    _api.invalidatedCursorArea.right = std::max(_api.invalidatedCursorArea.right, right);
+    _api.invalidatedCursorArea.bottom = std::max(_api.invalidatedCursorArea.bottom, bottom);
+    return S_OK;
+}
+
+[[nodiscard]] HRESULT AtlasEngine::InvalidateSystem(const RECT* const prcDirtyClient) noexcept
+{
+    const auto top = prcDirtyClient->top / _api.fontMetrics.cellSize.y;
+    const auto bottom = prcDirtyClient->bottom / _api.fontMetrics.cellSize.y;
+
+    // BeginPaint() protects against invalid out of bounds numbers.
+    SMALL_RECT rect;
+    rect.Top = gsl::narrow_cast<SHORT>(top);
+    rect.Bottom = gsl::narrow_cast<SHORT>(bottom);
+    return Invalidate(&rect);
+}
+
+[[nodiscard]] HRESULT AtlasEngine::InvalidateSelection(const std::vector<SMALL_RECT>& rectangles) noexcept
+{
+    for (const auto& rect : rectangles)
+    {
+        // BeginPaint() protects against invalid out of bounds numbers.
+        // TODO: rect can contain invalid out of bounds coordinates when the selection is being

```

Comment:
>// TODO: [](http://example.com/codeflow?start=8&length=8)

I thought you figured this out? Not yet?

Reply:
I couldn't... This will be my EIM thingy. 🙂